Week2 [STEP 3] Hisop

Hi-sop opened the pull request:
@juun97
STEP3 PR

## 고민했던 점

1. unwrap을 위해 guard-let을 사용할지 if-let을 사용할지 고민 -> 
  guard-let쪽이 조금 더 보기 좋은거같아 사용했습니다.

2. 기능별로 최대한 분리하여 함수로 나눠보고자 했습니다.
  로또 번호를 만들때 호출하는 함수는 makeLottoNumber이며 함수 내부에서 makeRound와 makeNumber를 호출해 전달받은 dictionary에 값을 넣도록 했습니다.
  dictionary를 체크하는 checkNumbers와 step2의 내용인 compareNumbers를 만들었습니다.

3. 여러번 함수를 호출해야하며 동일한 dictionary에 값을 추가해야 했으므로 Call by reference 로 주소를 넘겨주었습니다.

4. 저번 PR때 말씀해주신 if-else구문의 컨벤션을 수정했습니다.

Commit:
Feat: Step 요구사항 구현
Dictionary 타입을 사용하여 당첨번호 회차별로 기록
당첨번호 생성이 잘 이루어졌는지 checkNumbers로 확인
step2의 기능은 compareNumbers 함수로 변경하여 유지
기능별로 최대한 분리하여 함수로 나눔

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
+    return "\(round)회차"
+}
+
+func makeNumber() -> Set<Int> {
```
좀 더 명확한 네이밍이 필요해보입니다 🤔
예를들면, 어떻게 Number를 만드는지 처럼요 !

Hi-sop replied:
Random한 Number를 뽑아 만든다는 뜻으로 generateRandomNumbers로 변경했습니다.

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
+}
+
+func makeNumber() -> Set<Int> {
     var numbers: Set<Int> = []
-    
     repeat {
         let random = Int.random(in: 1...45)
         numbers.insert(random)
         } while numbers.count < 6
     return numbers
 }
```
코드의 개행부분은 Hisop님만의 기준이 있을까요?
현재는 개행이 존재하지 않아 가독성부분에 있어서 잘 안읽히는 느낌을 받습니다 😢

Hi-sop replied:
평소엔 막연하게 변수 선언부와 실행하는 부분을 나누는 방식으로 작업했던거같아요.
13번째줄은 수정중에 지워버린거같습니다..
명확한 컨벤션이 있으면 알려주실수 있을까요?

juun97 replied:
저 같은 경우 같은 맥락끼리는 개행을 없이 붙여씁니다
예를 들면 변수의 선언들이나 로직들의 사용들 처럼요. 추가적으로 return 전에는 개행을 추가해주곤 합니다 😃

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
+}
+checkNumbers(roundByNumbers: roundByNumbers)
+//step2
+let myLottoNumbers: [Int] = [1, 6, 12, 15, 23, 45]
```
숫자를 직접 할당하는게 아닌 저희가 만들어둔 함수를 활용해볼까요?

juun97 replied:
추가적으로 필요하지 않는 주석은 삭제를 하도록 할까요?

Hi-sop replied:
generateRandomNumbers를 받아오는것으로 변경했습니다.
타입은 Set<Int>로 변경했습니다. 비교한 후 출력하는 방식이라 당장 Int배열로 바꿔줄 필요가 없었습니다.

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
+var roundByNumbers: [String: Set<Int>] = [:]
+for _ in 1...5 {
+    makeLottoNumber(roundByNumbers: &roundByNumbers)
+}
```
함수로 한번 더 추상화가 가능할 거 같네요 👍🏻
어떻게 하면 우리가 원하는 라운드만큼의 로또넘버를 생성할 수 있을까요?

Hi-sop replied:
maxRound를 파라미터로 받아 해당 함수 내에서 반복문을 시행해 만들도록 변경했습니다. 
추가적으로 Set의 개수를 세어 라운드String을 만들던 방식에서 현재 iterator를 넘겨주어 라운드를 만들도록 했습니다.

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
-let result = winningNumber.intersection(myLottoNumbers).sorted().map { String($0) }.joined(separator:", ")
-if result.isEmpty {
-    print("아쉽지만 겹치는 번호가 없습니다.")
+func makeLottoNumber(roundByNumbers: inout [String: Set<Int>]) {
+    let round = makeRound(round: roundByNumbers.count + 1)
+    roundByNumbers[round] = makeNumber()
 }
```
위의 makeNumber 와 네이밍이 상당히 유사하네요! 네이밍으로 각 함수의 특징을 보여주면 더 좋을 것 같습니다 😃

Hisop님이 생각하셨을 때 현재 이 함수는 어떤 역할을 하고 있나요?

Hi-sop replied:
max라운드만큼 반복하며, 해당 라운드에 맞는 String을 만들어 맞는 key에 generateRandomNumbers에서 받은 리턴값을 넣는것으로 변경했습니다.
네이밍을 추가적으로 변경할 필요가 있을 것 같다면 알려주세요!

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
-else {
-    print("축하합니다! 겹치는 번호는 \(result) 입니다.")
+
+func checkNumbers(roundByNumbers: [String: Set<Int>]) {
+    print(roundByNumbers) //check Dictionary
+    for num in 1...6 {
+        let round = "\(num)회차"
+        guard let numbers = roundByNumbers[round] else {
+            break
+        }
+        let sortNumbers = numbers.sorted().map{ String($0) }
+        let numbersAddSeparator = sortNumbers.joined(separator: ", ")
+        print("\(round)의 로또 당첨 번호는 \(numbersAddSeparator) 입니다.")
+    }
 }
```
STEP3 의 명세사항에 보시면
> Dictionary 에서 추출한 원하는 회차의 로또 당첨 번호 6개를 아래와 같이 출력합니다.

라는 항목이 있습니다.

현재 코드는 특정회차가 아닌 모든 회차에 대한 정보를 호출 하고있네요 🤔
요구사항에 맞춰서 수정을 해볼까요?

Hi-sop replied:
코드 변경했습니다.

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
+    }
 }
+
+func compareNumbers(roundByNumbers: [String: Set<Int>], round num: Int, myLottoNumbers: [Int]) {
+    let round = "\(num)회차"
```
전달인자 레이블을 잘 활용 해주셨네요 👍🏻
하지만 Swift API Guideline 에 따르면 네이밍에서 축약형을 지양하라고 나와있습니다 🤔

Hi-sop replied:
forRound로 전달인자 레이블을 변경했습니다. num은 적절하지 않았던것같네요

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,22 +8,58 @@
+
+func compareNumbers(roundByNumbers: [String: Set<Int>], round num: Int, myLottoNumbers: [Int]) {
+    let round = "\(num)회차"
+    print("\(round)의 결과를 내 번호와 비교합니다")
+    
+    guard let numbers = roundByNumbers[round] else {
+        return
+    }
```
특정 회차가 존재하지 않는경우 compareNumbers 함수가 바로 종료되겠네요 🤔
그렇다면 그 결과에 대한 출력도 같이 진행해주시는게 좋아보입니다 😃

Hi-sop replied:
리턴 이전에 "\(roundString)는 진행되지 않았습니다." 를 출력하도록 했습니다.

juun97 reviewed (commented):
## 고민했던 점
> 1. unwrap을 위해 guard-let을 사용할지 if-let을 사용할지 고민 ->
guard-let쪽이 조금 더 보기 좋은거같아 사용했습니다.

guard let 구문을 활요하여 옵셔널 바인딩을 진행해주셨네요 👍🏻
if 문과 guard 문의 차이점과 어떤상황에서 주로 쓰이는지 알려주시겠어요?


> 2. 기능별로 최대한 분리하여 함수로 나눠보고자 했습니다.
로또 번호를 만들때 호출하는 함수는 makeLottoNumber이며 함수 내부에서 makeRound와 makeNumber를 호출해 전달받은 dictionary에 값을 넣도록 했습니다.
dictionary를 체크하는 checkNumbers와 step2의 내용인 compareNumbers를 만들었습니다.

매우 좋습니다 👍🏻
함수가 하나의 동작만 하도록 기능을 분리해준다면 함수의 네이밍 짓기도 더 수월하죠 😄

다른 장점들은 Hisop님이 알아보시고 알려주시겠어요?

> 3. 여러번 함수를 호출해야하며 동일한 dictionary에 값을 추가해야 했으므로 Call by reference 로 주소를 넘겨주었습니다.

inout 을 잘 활용해주셨네요 👍🏻
Dictionary 를 어떤 이유 때문에 주소를 넘겨 주었을까요?

추가로 지금 같은 상황처럼 다양한 곳에서 특정 객체를 접근해야 하는 상황이라면 어떤 방법을 사용할 수 있을까요?


> 4. 저번 PR때 말씀해주신 if-else구문의 컨벤션을 수정했습니다.

좋습니다 👍🏻

Commit:
Refactor: 피드백 내용 반영하여 리팩토링
makeNumber()를 generateRandomNumbers()로 변경
내 로또넘버를 직접 입력하지 않고 함수 호출해 받도록 변경
makeLottoNumber()의 기능 변경
parameter 이름 변경
checkNumbers는checkAllNumbers로 바꾼 후 주석처리, 요구사항대로 함수 변경

Hi-sop commented:
1. if 문과 guard 문의 차이점
  if와 guard는 같은 의미로 읽혀서 가독성과 취향의 차이가 아닌가 생각합니다.
  guard는 let을 코드블럭 밖에서도 사용이 가능하다는 특징이 있습니다.
  저는 조건문을 통해 해당 함수나 반복문을 종료하고 싶을 때 guard를 사용할듯 합니다.

2. 함수가 하나의 동작만 하도록 기능을 분리해줄 때 장점
  찾아보니 테스트에 용이하며 가독성과 재사용성을 챙길수 있다고 합니다.

3. Dictionary 를 어떤 이유 때문에 주소를 넘겨 주었을까요?
  Dictonary의 value를 넘겨주게 되면 제가 원하는 roundByNumbers에 값을 추가하는게 아닌 
  복사본에 값을 추가하게 되는 것이라 생각했습니다.

4. 추가로 지금 같은 상황처럼 다양한 곳에서 특정 객체를 접근해야 하는 상황이라면 어떤 방법을 사용할 수 있을까요?
  애초에 클래스로 만들어 value타입이 아닌 reference type을 사용해 볼 것 같습니다.
  
  다시 제 코드를 읽어보니 main.swift에 쓴 변수는 글로벌 스코프에 선언한거라 파라미터로 전달하지 않고 꺼내써도 문제되지 않을거같네요...?
 -> 반영하여 전달하지 않고 꺼내쓰도록 수정했습니다.

Commit:
Comment : 불필요한 주석 삭제
사실 checkAllNumbers도 필요 없지 않은가 싶
리턴 전에 개행 추가

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -12,22 +12,26 @@ func makeRound(round: Int) -> String {
 }
 
-func makeNumber() -> Set<Int> {
+func generateRandomNumbers() -> Set<Int> {
```
좀 더 명확한 네이밍이 되었네요 👍🏻

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -38,12 +42,26 @@ func checkNumbers(roundByNumbers: [String: Set<Int>]) {
-func compareNumbers(roundByNumbers: [String: Set<Int>], round num: Int, myLottoNumbers: [Int]) {
-    let round = "\(num)회차"
-    print("\(round)의 결과를 내 번호와 비교합니다")
+func compareNumbers(forRound round: Int) {
```
전에 코멘트 드렸던 내용중에 함수의 네이밍은 하나의 문장처럼 읽혀야 된다는 내용이 있었습니다.
파라미터의 네이밍도 마찬가지로 함수의 네이밍과 연결되어서 자연스럽게 읽혀야 합니다.

1. compareNumbers forRound round
2. compareNumber for round

어떤게 더 자연스럽게 느껴지시나요?

(추가적으로, 파라미터의 전달인자 레이블의 경우 전치사를 많이 사용하곤합니다)

Hi-sop replied:
2번이 더 자연스러워 보이네요

Hi-sop replied:
제 코드에선 호출할때는 compareNumbers(forRound: ?) 로 호출했으면 하고, 함수 내부적으로는 round로 사용하고 싶은 의도로 작성했습니다.

Hi-sop replied:
이런 경우에도
```swift
func compareNumbers(forRound: Int) {

}
```
로 네이밍해주는게 더 적합하다는 말씀이신거죠?

juun97 replied:
forRound가 아닌 for round 를 의미한거였습니다!

Hi-sop replied:
이해 못했어요.. 다시 설명해주실수있나요

juun97 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -12,22 +12,26 @@ func makeRound(round: Int) -> String {
-func makeLottoNumber(roundByNumbers: inout [String: Set<Int>]) {
-    let round = makeRound(round: roundByNumbers.count + 1)
-    roundByNumbers[round] = makeNumber()
+func makeLottoNumber(maxRound: Int) {
+    for currentRound in 1...maxRound {
+        let round = makeRound(round: currentRound)
+        
+        roundByNumbers[round] = generateRandomNumbers()
+    }
```
원하는 round 수만큼의 로또 번호를 생성하는것으로 변경되었네요 좋습니다 👍🏻

juun97 reviewed (commented):
>1. if 문과 guard 문의 차이점
if와 guard는 같은 의미로 읽혀서 가독성과 취향의 차이가 아닌가 생각합니다.
guard는 let을 코드블럭 밖에서도 사용이 가능하다는 특징이 있습니다.
저는 조건문을 통해 해당 함수나 반복문을 종료하고 싶을 때 guard를 사용할듯 합니다.

그렇습니다 큰 차이점으로는 scope 밖에서 사용할 수 있는지가 있습니다.
그 밖에도 guard 문은 빠른 종료를 목적으로 한다는 특징이 있습니다. guard 문을 사용하게 되면 조건이 참이어야만 다음 문장으로 넘어 갈 수 있기에 해당 함수를 한 번에 훑어 내려 볼 때 가독성면에서 if 문 보다 나은 점이 있습니다. 

> 2. 함수가 하나의 동작만 하도록 기능을 분리해줄 때 장점
찾아보니 테스트에 용이하며 가독성과 재사용성을 챙길수 있다고 합니다.

잘 알아봐주셨습니다 👍🏻

> 3. Dictionary 를 어떤 이유 때문에 주소를 넘겨 주었을까요?
Dictonary의 value를 넘겨주게 되면 제가 원하는 roundByNumbers에 값을 추가하는게 아닌 복사본에 값을 추가하게 되는 것이라 생각했습니다.

네 맞습니다 👍🏻
좀 더 자세히 알아보자면 Dictionary 의 경우 struct 로 구현되어있고 struct 는 Value 타입 입니다. 그리고 Swift 에서 함수의 파라미터는 상수이므로 두 조건이 맞물리게 되면서 함수 내부에서 파라미터 값을 변경해도 반영이 안됩니다.

> 4. 추가로 지금 같은 상황처럼 다양한 곳에서 특정 객체를 접근해야 하는 상황이라면 어떤 방법을 사용할 수 있을까요?
애초에 클래스로 만들어 value타입이 아닌 reference type을 사용해 볼 것 같습니다.

네 reference type 을 사용하는 것도 하나의 방법이죠 👍🏻
그런데 지금 같은 상황에서는 파일의 전역에서 공유할 수 있는 전역변수의 사용을 고려해봐도 좋을 것 같습니다 😃

> 다시 제 코드를 읽어보니 main.swift에 쓴 변수는 글로벌 스코프에 선언한거라 파라미터로 전달하지 않고 꺼내써도 문제되지 않을거같네요...?
-> 반영하여 전달하지 않고 꺼내쓰도록 수정했습니다.

네 현재 수정하신 방향이 전역변수를 활용하신 사례입니다 👍🏻
그렇다면 전역변수를 사용할 때 어떤 점을 조심해야하는지 알려주시겠어요?

Hi-sop commented:
전역변수는 생명주기가 길고 어느 함수에서나 접근이 가능하기 때문에
의도치 않은 값 변경에 주의해야할 것 같습니다.
또 전역변수와 함수 내부의 변수 이름이 같다면 문제가 생길수도 있을거같습니다.
